dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -1003,6 +1027,29 @@ DDS::ReturnCode_t DynamicData::get_wstring_value(ACE_CDR::WChar*& value, MemberI
   return get_single_value<TK_STRING16>(value, id);
 }
 
+DDS::ReturnCode_t DynamicData::get_discriminator_value(DynamicData& value) {
```

Comment:
I don't think we need a separate method for reading discriminator as it can be read using the existing interfaces. Something like the one Adam suggested in chat would work, but I think it can be simpler if we use a sentinel ID value, like MEMBER_ID_INVALID, to indicate the discriminator. For example, if the discriminator type is `unsigned short`, its value can be read using `union_data.get_uint16_value(disc_value, MEMBER_ID_INVALID)` where `union_data` is a `DynamicData` object of the union. This would require a bit of change to `get_value_from_union` in `DynamicData` but not a new method.

Reply:
How about using a value that's above MEMBER_ID_MAX but is not MEMBER_ID_INVALID?
As in the code I posted before, this shouldn't be a detail the user needs to know.  It would be returned by `get_member_id_by_name`.

Reply:
I think this is one of the uses of MEMBER_ID_INVALID as stated in the spec, where the ID is missing or irrelevant. Even though the spec doesn't say explicitly, if we let users call `get_*_value` with MEMBER_ID_INVALID, `DynamicData` can decide that the user wants to read the dicsriminator without needing to delegate to `DynamicType` and related classes. Else if we choose to delegate to `DynamicType`-related classes, they would need to change too because right now discriminator is not a member in the member list in `DynamicType` for union.

Reply:
I'm not sure `DynamicType` would need to change in order for `DynamicData` to treat this differently.

I suggest not using MEMBER_ID_INVALID since that would make it seem like `get_member_id_by_name` failed.

Reply:
`get_member_id_by-name` isn't being used in this case and it would fail regardless because the discriminator is not a DynamicTypeMember of the DynamicType of the union.